hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -881,4 +894,117 @@ private static RetryTemplate createRetryTemplate() {
+                } finally {
+                    lock.unlock();
+                }
+            } catch (final InterruptedException e) {
+                LOG.error("InterruptedException while invalidating distribution sets {}!",
+                        distributionSetInvalidation.getDistributionSetIds(), e);
+                Thread.currentThread().interrupt();
+            }
+        } else {
+            // no lock is needed as no rollout will be stopped
+            invalidateDistributionSetsInTransaction(distributionSetInvalidation, tenant);
+        }
+    }
+
+    private void invalidateDistributionSetsInTransaction(final DistributionSetInvalidation distributionSetInvalidation,
+            final String tenant) {
+        DeploymentHelper.runInNewTransaction(txManager, tenant + "-invalidateDS", status -> {
+            distributionSetInvalidation.getDistributionSetIds().forEach(setId -> invalidateDistributionSet(setId,
+                    distributionSetInvalidation.getCancelationType(), distributionSetInvalidation.isCancelRollouts()));
+            return 0;
+        });
+    }
+
+    private void invalidateDistributionSet(final long setId, final CancelationType cancelationType,
+            final boolean cancelRollouts) {
+        final JpaDistributionSet set = (JpaDistributionSet) distributionSetManagement.getValidAndComplete(setId);
+        set.invalidate();
+        distributionSetRepository.save(set);
+        LOG.debug("Distribution set {} set to invalid", set.getId());
+
+        if (shouldRolloutsBeCanceled(cancelationType, cancelRollouts)) {
+            cancelRollouts(set);
+        }
+
+        if (cancelationType != CancelationType.NONE) {
+            cancelActions(cancelationType, set);
+        }
+
+        cancelAutoAssignments(setId);
+    }
+
+    private boolean shouldRolloutsBeCanceled(final CancelationType cancelationType, final boolean cancelRollouts) {
+        return cancelationType != CancelationType.NONE || cancelRollouts;
+    }
+
+    private void cancelRollouts(final DistributionSet set) {
+
+        // stop all rollouts for this distribution set
+        rolloutRepository.findByDistributionSetAndStatusIn(set, ROLLOUT_STATUS_STOPPABLE).forEach(rollout -> {
+            final JpaRollout jpaRollout = (JpaRollout) rollout;
+            jpaRollout.setStatus(RolloutStatus.STOPPING);
+            rolloutRepository.save(jpaRollout);
+            LOG.debug("Rollout {} stopped", jpaRollout.getId());
+        });
+
+    }
+
+    private void cancelActions(final CancelationType cancelationType, final DistributionSet set) {
+        actionRepository.findByDistributionSetAndActiveIsTrue(set).forEach(action -> {
+            final JpaAction jpaAction = (JpaAction) action;
+            cancelAction(jpaAction.getId());
+            LOG.debug("Action {} canceled", jpaAction.getId());
+        });
+        if (cancelationType == CancelationType.FORCE) {
```

Comment:
please check if it is possible to force quite running action by Mgmt Api without canceling it beforehand

Reply:
This is not possible, also via REST API you get "10202 is not canceled yet and cannot be force quit" as a response in this case